bin/node/runtime/src/lib.rs

```diff
@@ -468,11 +468,102 @@ parameter_types! {
 	pub const SlashDeferDuration: pallet_staking::EraIndex = 24 * 7; // 1/4 the bonding duration.
 	pub const RewardCurve: &'static PiecewiseLinear<'static> = &REWARD_CURVE;
 	pub const MaxNominatorRewardedPerValidator: u32 = 256;
-}
+	// enable the below flag to allow dynamic damping of validator count per era.
+	pub const EnableAutomaticValidatorUpdatePerEra: bool = true;
+	pub const BottomXPercentOfValidators: u32 = 20;
+	pub const BottomYPercentOfValidators: u32 = 40;
```

Comment:
```suggestion
	pub const BottomXPercentOfValidators: u32 = 20;
	pub const BottomYPercentOfValidators: u32 = 40;
```

These two should be hardcoded in the implementation. No need to pass them into staking. As recommended, the correct API is `Convert<(u32, Vec<BalanceOf<T>>), u32>`. You can read `max_validator_count` as `<pallet_staking::ValidatorCount<Runtime>>::get()`. And the last two should just NOT be in the api in any way. Instead, they can and should be parameters of `AutomaticValidatorUpdatePerEra `.

```
AutomaticValidatorUpdatePerEra<X: Get<u32>, Y: Get<u32>>( /* need to work with std::marker::PhantomData<(Xy, Y)> */)

impl<X: .., Y: ..> Convert<(u32, Vec<BalanceOf<T>>), u32> for AutomaticValidatorUpdatePerEra<X, Y> {
  let x = X::get();
  let y = Y::get();
}
```

Then any top-level runtime can configure these, but they are not exposed to staking internally, which exactly the sensible separation of concerns here.